### PR TITLE
Evaluation Methods Documentation

### PR DESCRIPTION
**Pull request by czaloom**

### CODE REVIEW

File: docs/metrics.md [hunk 1]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
-| Name 	| Description 	|
-|---	|---	|
-| Average Precision (AP) 	| The weighted mean of precisions achieved at several different recall thresholds for a single Intersection over Union (IOU)*, grouped by class. |
-| AP Averaged Over IOUs 	| The average of several AP metrics, calculated at various IOUs, grouped by class. 	|
-| Mean Average Precision (mAP) 	| The mean of several AP scores, calculated over various classes.	|
-| mAP Averaged Over IOUs 	| The mean of several averaged AP scores, calculated over various classes. 	|
+| Name | Description | Equation |
+| :- | :- | :- |
+| Average Precision (AP) | The weighted mean of precisions achieved at several different recall thresholds for a single Intersection over Union (IoU), grouped by class. | See [AP methods](#average-precision-ap-for-object-detection). |
+| AP Averaged Over IoUs | The average of several AP metrics, calculated at various IoUs, grouped by class. | $$\dfrac{1}{\|thresholds\|} \sum\limits_{iou \in thresholds} AP_{iou}$$ |
```
Comment (ntlind) [inline docs/metrics.md:22]: can we change "thresholds" to "number of IOU thresholds"?

can we make sure IOU is capitalized everywhere?

File: docs/metrics.md [hunk 2]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+| Mean Average Precision (mAP) 	| The mean of several AP scores, calculated over various classes. | $$\dfrac{1}{\|classes\|} \sum\limits_{c \in classes} AP_{c}$$ |
```
Comment (ntlind) [inline docs/metrics.md:23]: classes -> number of classes?

File: docs/metrics.md [hunk 3]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+| mAP Averaged Over IoUs | The mean of several averaged AP scores, calculated over various classes. | $$\dfrac{1}{\|thresholds\|} \sum\limits_{iou \in thresholds} mAP_{iou}$$ |
```
Comment (ntlind) [inline docs/metrics.md:24]: thresholds -> number of thresholds

File: docs/metrics.md [hunk 4]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
 ## Semantic Segmentation Metrics
 
-| Name 	| Description 	|
-|---	|---	|
-| Intersection Over Union (IOU) 	| The overlap between the ground truth and predicted regions of an image, measured as a percentage, grouped by class. IOUs are calculated by a) fetching the ground truth and prediction rasters for a particular image and class, b) counting the true positive pixels (e.g., the number of pixels that were selected in both the ground truth masks and prediction masks), and c) dividing the sum of true positives by the total number of pixels in both the ground truth and prediction masks. |
-| Mean IOU 	| The average of IOUs, calculated over several different classes. 	|
+| Name | Description | Equation |
+| :- | :- | :- |
+| Intersection Over Union (IoU) | A ratio between the groundtruth and predicted regions of an image, measured as a percentage, grouped by class. |$$\dfrac{area( prediction \cap groundtruth )}{area( prediction \cup groundtruth )}$$ |
+| Mean IoU 	| The average of IoUs, calculated over several different classes. | $$\dfrac{1}{\|classes\|} \sum\limits_{c \in classes} IoU_{c}$$ |
```
Comment (ntlind) [inline docs/metrics.md:31]: "classes" in divisor -> "number/count of classes"?

File: docs/metrics.md [hunk 5]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+# Methods
```
Comment (ntlind) [inline docs/metrics.md:33]: you're grouping these "methods" by metrics, so this naming seems odd. what about calling this "Detailed Metric Calculations" or "Appendix: Detailed Metric Calculations"?

File: docs/metrics.md [hunk 6]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+## ROC AUC for Classification
+
+### Determining Binary Truth
```
Comment (ntlind) [inline docs/metrics.md:37]: nit: is "binary truth" the correct phrasing here? it reads awkwardly, but maybe it's a technical term I'm not aware of

File: docs/metrics.md (hunk 1 above)
Comment (czaloom) [reply docs/metrics.md:22]: IoU is the correct format i believe

File: docs/metrics.md [hunk 7]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+### Area under the ROC curve (ROC AUC)
+
+From the ROC curve we calculate the ROC AUC metric by taking the integral using the trapezoidal rule formula.
+
+$$
+\sum_{i=1}^{|scores|} \frac{(TPR(scores_{i-1}) - TPR(scores_i)) \cdot (FPR(scores_{i-1}) - FPR(scores_i))}{2}
+$$
+
+### References
+- [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc)
```
Comment (ntlind) [inline docs/metrics.md:63]: nit: can we delete this reference section since there's only one reference, and hyperlink to this page somewhere else in your above explaination?

File: docs/metrics.md (hunk 2 above)
Comment (czaloom) [reply docs/metrics.md:23]: the vertical bars signify a quantity

File: docs/metrics.md (hunk 3 above)
Comment (czaloom) [reply docs/metrics.md:24]: the vertical bars signify a quantity

File: docs/metrics.md (hunk 4 above)
Comment (czaloom) [reply docs/metrics.md:31]: the vertical bars signify a quantity

File: docs/metrics.md [hunk 8]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+## Average Precision (AP) for Object Detection
+
+For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IoU) of geometric annotations.
+
+### What is Intersection-over-Union (IoU)?
+
+IoU is the ratio of intersecting area over total area spanned by two geometries.
```
Comment (ntlind) [inline docs/metrics.md:71]: capitalize IOU everywhere for consistency

File: docs/metrics.md [hunk 9]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+- $\text{False Positive Rate (FPR)} = \dfrac{|FP|}{|FP| + |TN|}$
+
+### Receiver Operating Characteristic (ROC)
+
+An ROC curve plots TPR vs. FPR at different confidence thresholds.
+
+### Area under the ROC curve (ROC AUC)
+
+From the ROC curve we calculate the ROC AUC metric by taking the integral using the trapezoidal rule formula.
```
Comment (ntlind) [inline docs/metrics.md:56]: "taking the integral" -> "estimating the integral"?

File: docs/metrics.md [hunk 10]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+$$
+\sum_{i=1}^{|scores|} \frac{(TPR(scores_{i-1}) - TPR(scores_i)) \cdot (FPR(scores_{i-1}) - FPR(scores_i))}{2}
```
Comment (ntlind) [inline docs/metrics.md:59]: did you pull this formula from somewhere? it looks a little off to me, but I can't quite put my finger on why. I'd be interested to check my understanding by reading your source

File: docs/metrics.md [hunk 11]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+$$
+
+### References
+- [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc)
+
+## Average Precision (AP) for Object Detection
```
Comment (ntlind) [inline docs/metrics.md:65]: remove "for Object Detection" since you're also talking about instance segmentation?

File: docs/metrics.md [hunk 12]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+# Methods
+
+## ROC AUC for Classification
+
+### Determining Binary Truth
+
+| Element | Description |
+| ------- | ------------ |
+| True Positive (TP) | Prediction confidence is above a threshold and is correct. |
```
[the remark this answers is not in the document]
Comment (czaloom) [reply docs/metrics.md:41]: Classification doesnt use Iou

File: docs/metrics.md (hunk 8 above)
Comment (czaloom) [reply docs/metrics.md:71]: Im fairly sure IoU is correct

Comment (ntlind) [inline docs/metrics.md:71]: phrasing here reads a little awkwardly. what do you think about:

IOU is the ratio of the intersecting area of two geometries divided by the total area of those same geometries.

File: docs/metrics.md (hunk 10 above)
Comment (czaloom) [reply docs/metrics.md:59]: Its the trapezoidal rule formula modified for our use

Comment (czaloom) [reply docs/metrics.md:59]: Ill work on it

File: docs/metrics.md [hunk 13]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+IoU is the ratio of intersecting area over total area spanned by two geometries.
+
+$$Intersection \ over \ Union \ (IoU) = \dfrac{Area( prediction \cap groundtruth )}{Area( prediction \cup groundtruth )}$$
+
+### Multiclass Precision and Recall
+
+| Case | Description |
+| :- | :- |
+| True Positive (TP) | Prediction meets IoU threshold requirements. |
```
Comment (ntlind) [inline docs/metrics.md:79]: "Prediction meets IoU threshold requirements." - can you clarify what this means?

File: docs/metrics.md (hunk 8 above)
Comment (czaloom) [reply docs/metrics.md:71]: IoU is the ratio of intersecting area over the joint area spanned by the two geometries.

File: docs/metrics.md [hunk 14]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+| True Negative (TN) | Unused in multi-class evaluation.
+| False Negative (FN) | No prediction exists. |
+
+- $Precision = \dfrac{|TP|}{|TP| + |FP|} = \dfrac{\text{Number of True Predictions}}{|\text{Predictions}|}$
+
+- $Recall = \dfrac{|TP|}{|TP| + |FN|} = \dfrac{\text{Number of True Predictions}}{|\text{Groundtruths}|}$
+
+### Finding the best prediction for a groundtruth.
+
+To properly evaluate a detection we must first find the best matches of predictions to ground truths. We start by iterating over predictions by score from highest to lowest. For each prediction we assign the ground truth with the highest IoU value. Both the prediction and ground truth are now considered paired and removed from the pool of choices.
```
Comment (ntlind) [inline docs/metrics.md:90]: - comma after "detection"
- iterating over predictions by score from highest to lowest -> iterating over our predictions, starting with the highest scores first
- For each prediction we assign the ground truth with the highest IoU value. -> We pair each prediction with the ground truth that has the highest calculated IOU.

File: docs/metrics.md [hunk 15]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+Note: For simplicity, the following algorithm assumes operation over the same datum and label.
+
+$$
+\begin{aligned}
+&\underline{\large\textbf{Algorithm 1} \hspace{0.5em} \text{Rank IOU's}} \\
```
Comment (ntlind) [inline docs/metrics.md:96]: I find this latex very confusing; I wonder if you could drastically shorten/simplify this section. some specific items I'm confused by:
- you offer more "Notes"s and summary details in the latex, rather than just including them in the preceeding text
- you're assigning variables rather than just using full names (like you did in the equations for previous sections)
- you're mixing python nomenclature (argsort) with (what I think is) pseudocode (e.g., foreach; for x do...); why not just make it as pythonic as possible without using variable names given we use python in our other examples?

File: docs/metrics.md [hunk 16]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+&23 \hspace{1em} \textbf{end} \\
+\end{aligned}
+$$
+
+### Precision-Recall Curve
+
+We can now compute the precision-recall curve using our previously ranked IoU's. We do this by iterating through the ranked IoU's and creating points cumulatively using recall and precision.
+
+$$
+\begin{aligned}
```
Comment (ntlind) [inline docs/metrics.md:135]: same comments above apply here

File: docs/metrics.md [hunk 17]
```diff
@@ -4,30 +4,173 @@ Let's look at the various metrics you can calculate using Valor.
...
+\rho_{interp} = \underset{\tilde{r}:\tilde{r} \ge r}{max \ \rho (\tilde{r})}
+$$
+
+### References
+- [MS COCO Detection Evaluation](https://cocodataset.org/#detection-eval)
+- [Mean Average Precision (mAP) Using the COCO Evaluator](https://pyimagesearch.com/2022/05/02/mean-average-precision-map-using-the-coco-evaluator/)
 
-\*When calculating IOUs for object detection metrics, Valor handles the necessary conversion between different types of image annotations. For example, if your model prediction is a polygon and your ground truth is a raster, then the raster will be converted to a polygon prior to calculating the IOU.
\ No newline at end of file
+### Notes
```
Comment (ntlind) [inline docs/metrics.md:175]: can you move this up to below the original "Object Detection" section now that you've added more content between the two?

File: docs/metrics.md (hunk 11 above)
Comment (ntlind) [reply docs/metrics.md:65]: overall, this section was a bit hard to follow. could we add a signpost here at the top describing the main steps, then use the steps as the headers?

### Step 1: Use IOU to Match Predictions with Groundtruths
...

File: docs/metrics.md (hunk 4 above)
Comment (ntlind) [reply docs/metrics.md:31]: see above

File: docs/metrics.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ntlind) [reply docs/metrics.md:41]: you're right. clarify what you mean by "threshold" here

File: docs/metrics.md (hunk 8 above)
Comment (ntlind) [reply docs/metrics.md:71]: see above

File: docs/metrics.md (hunk 9 above)
Comment (ntlind) [reply docs/metrics.md]: you aren't literally taking the integral in the equation, so you should delete that reference or change it to "estimating"

File: docs/metrics.md [hunk 18]
```diff
@@ -4,30 +4,189 @@ Let's look at the various metrics you can calculate using Valor.
...
+$$
+ROC AUC =  \sum_{i=1}^{|scores|} \frac{  \lVert Point(score_{i-1}) - Point(score_i) \rVert }{2}
+$$
+
+### References
+- [Classification: ROC Curve and AUC](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc)
+
+## Average Precision (AP)
+
+For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IoU) of geometric annotations.
```
Comment (ekorman) [inline docs/metrics.md:86]: i prefer replacing "annotations" with "predictions and groundtruths". 

people usually don't think of an "annotation" as a prediction

File: docs/metrics.md [hunk 19]
```diff
@@ -4,30 +4,189 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+### Multiclass Precision and Recall
+
+Geometric annotations use the ratio intersection-over-union (IoU) to calculate precision and recall. IoU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.
```
Comment (ekorman) [inline docs/metrics.md:90]:
```suggestion
Tasks that predict geometries (such as object detection or instance segmentation) use the ratio intersection-over-union (IoU) to calculate precision and recall. IoU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.
```

File: docs/metrics.md (hunk 1 above)
Comment (czaloom) [reply docs/metrics.md:22]: Fixed!

File: docs/metrics.md (hunk 2 above)
Comment (ekorman) [reply docs/metrics.md:23]: i'm familiar with vertical bars as cardinality but can't hurt to define it once before using in this way

File: docs/metrics.md (hunk 7 above)
Comment (ekorman) [reply docs/metrics.md:82]: i agree if there's just one link in reference section than the section is unnecessary

File: docs/metrics.md [hunk 20]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+$$
+TPR(threshold) = \dfrac{|TP(threshold)|}{|TP(threshold)| + |FN(threshold)|}
+$$
+
+$$
+FPR(threshold) = \dfrac{|FP(threshold)|}{|FP(threshold)| + |TN(threshold)|}
+$$
+
+In Valor, we use the confidence scores sorted in decreasing order as our thresholds.
```
Comment (ntlind) [inline docs/metrics.md:69]: I was confused by what "threshold" meant in the above calculations until I read this sentence. maybe move this sentence up under the header, and change it to: 

"In Valor, we use the confidence scores sorted in decreasing order as our thresholds. Using these thresholds, we can calculate our TPR and FPR as follows:"

File: docs/metrics.md [hunk 21]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+$$
+Point(score) = (FPR(score), \ TPR(score))
+$$
+
+### Area under the ROC curve (ROC AUC)
```
Comment (ntlind) [inline docs/metrics.md:75]: capitalize the first letters in this header to be consistent with other headers?

File: docs/metrics.md [hunk 22]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
+$$Intersection \ over \ Union \ (IOU) = \dfrac{Area( prediction \cap groundtruth )}{Area( prediction \cup groundtruth )}$$
+
+By setting a threshold we can use IOU to determine whether we count a prediction-groundtruth pairing based on their overlap.
+
+| Case | Description |
+| :- | :- |
+| True Positive (TP) | Prediction-GroundTruth pair exists with IOU >= threshold. |
+| False Positive (FP) | Prediction-GroundTruth pair exists with IOU < threshold. |
+| True Negative (TN) | Unused in multi-class evaluation.
+| False Negative (FN) | No Prediction exists for the GroundTruth. |
```
Comment (ntlind) [inline docs/metrics.md:100]:
```suggestion
| False Negative (FN) | No Prediction with a particular label exists for the GroundTruth. |
```

File: docs/metrics.md [hunk 23]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+For object-detection and instance segmentation tasks, average-precision is calculated from the intersection-over-union (IOU) of geometric predictions and ground truths.
+
+### Multiclass Precision and Recall
+
+Tasks that predict geometries (such as object detection or instance segmentation) use the ratio intersection-over-union (IOU) to calculate precision and recall. IOU is the ratio of intersecting area over the joint area spanned by the two geometries and is defined in the following equation.
+
+$$Intersection \ over \ Union \ (IOU) = \dfrac{Area( prediction \cap groundtruth )}{Area( prediction \cup groundtruth )}$$
+
+By setting a threshold we can use IOU to determine whether we count a prediction-groundtruth pairing based on their overlap.
```
[the remark this answers is not in the document]
Comment (ntlind) [reply docs/metrics.md:93]: using "GroundTruth" to reference the class is fine

File: docs/metrics.md [hunk 24]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
+        if ranked_ious[i] >= threshold:
+            count_tp += 1
+        precision = count_tp / (i + 1)
+        recall = count_tp / number_of_groundtruths
+        retval.append((recall, precision))
+```
+
+### Calculating Average Precision
+
+Average precision is defined as the integration of the precision-recall curve. However, due to the varying nature of datasets it has been shown that interpolating this curve with a fixed set of points helps to reduce inconsistencies between dataset splits. The defacto standard has been to use a 101-point interpolation of the precision-recall curve to compute this integral.
```
[the remark this answers is not in the document]
Comment (ntlind) [reply docs/metrics.md:149]: "However, due to the varying nature of datasets " can you clarify what this means?

File: docs/metrics.md [hunk 25]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
-| Mean IOU 	| The average of IOUs, calculated over several different classes. 	|
+| Name | Description | Equation |
+| :- | :- | :- |
+| Intersection Over Union (IOU) | A ratio between the groundtruth and predicted regions of an image, measured as a percentage, grouped by class. |$$\dfrac{area( prediction \cap groundtruth )}{area( prediction \cup groundtruth )}$$ |
+| Mean IOU 	| The average of IOUs, calculated over several different classes. | $$\dfrac{1}{\text{number of classes}} \sum\limits_{c \in classes} IOU_{c}$$ |
+
+# Notes
+1. When calculating IOUs for object detection metrics, Valor handles the necessary conversion between different types of geometric annotations. For example, if your model prediction is a polygon and your groundtruth is a raster, then the raster will be converted to a polygon prior to calculating the IOU.
+
+# Appendix: Metric Calculations
```
Comment (ntlind) [inline docs/metrics.md:36]: overall I really like this new flow, appreciate you making these changes!

**Review by ntlind — CHANGES_REQUESTED** [review]

File: docs/metrics.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (czaloom) [reply docs/metrics.md:149]: Rewrote this section

File: docs/metrics.md [hunk 26]
```diff
@@ -4,30 +4,158 @@ Let's look at the various metrics you can calculate using Valor.
...
-| Name 	| Description 	|
-|---	|---	|
-| Intersection Over Union (IOU) 	| The overlap between the ground truth and predicted regions of an image, measured as a percentage, grouped by class. IOUs are calculated by a) fetching the ground truth and prediction rasters for a particular image and class, b) counting the true positive pixels (e.g., the number of pixels that were selected in both the ground truth masks and prediction masks), and c) dividing the sum of true positives by the total number of pixels in both the ground truth and prediction masks. |
-| Mean IOU 	| The average of IOUs, calculated over several different classes. 	|
+| Name | Description | Equation |
+| :- | :- | :- |
+| Intersection Over Union (IOU) | A ratio between the groundtruth and predicted regions of an image, measured as a percentage, grouped by class. |$$\dfrac{area( prediction \cap groundtruth )}{area( prediction \cup groundtruth )}$$ |
+| Mean IOU 	| The average of IOUs, calculated over several different classes. | $$\dfrac{1}{\text{number of classes}} \sum\limits_{c \in classes} IOU_{c}$$ |
+
+# Notes
```
[the remark this answers is not in the document]
Comment (czaloom) [reply docs/metrics.md:33]: Trying out github's markdown footnote feature

File: docs/metrics.md [hunk 27]
```diff
@@ -4,30 +4,162 @@ Let's look at the various metrics you can calculate using Valor.
...
+
+$$
+TPR(threshold) = \dfrac{|TP(threshold)|}{|TP(threshold)| + |FN(threshold)|}
+$$
+
+$$
+FPR(threshold) = \dfrac{|FP(threshold)|}{|FP(threshold)| + |TN(threshold)|}
+$$
+
+We now use the confidence scores sorted in decreasing order as our thresholds to generate points on a curve.
```
Comment (ntlind) [inline docs/metrics.md:68]:
```suggestion
We now use the confidence scores, sorted in decreasing order, as our thresholds in order to generate points on a curve.
```

**Review by ntlind — CHANGES_REQUESTED** [review]

**Review by ntlind — APPROVED** [review]
